enhancements/update/dont-require-registry-during-boot-and-upgrade.md

```diff
@@ -0,0 +1,201 @@
+
+## Summary
+
+Ensure that clusters don't require a registry server to boot or upgrade when
+all the required images have already been pulled.
+
+## Motivation
+
+Currently during reboots and upgrades clusters may need to contact the image
+registry servers, even if the images have already been pulled. This complicates
+things for clusters that are completely disconnected or that have an slow or
+unreliable connection to the image registry servers.
+
+### User Stories
+
+#### Boot without registry
+
+As the administrator of a cluster that has all the required images already
+pulled in all the nodes, I want to be able to reboot it without requiring
+access to a registry server.
+
+### Upgrade without registry
+
+As the administrator of a cluster that has all the required images already
+pulled in all the nodes, I want to be able to upgrade it without requiring
+access to a registry server.
+
+### Goals
+
+Ensure that clusters don't require a registry server to boot or upgrade when
+all the required images have already been pulled.
+
+### Non-Goals
+
+It is not the goal of this enhancement to add a mechanism to ensure that
+required images are available. That is the subject of the [pin and pre-load
+images](https://github.com/openshift/enhancements/pull/1481) enhancement.
+
+## Proposal
+
+### Workflow Description
+
+1. The administrator of a cluster boots a node or performs an upgrade.
+
+1. All the components of the cluster ensure that if the required images have
+been already pulled they will not try to contact the registry server.
+
+### API Extensions
+
+None.
+
+### Implementation Details/Notes/Constraints
+
+#### Don't use the `Always` pull policy
+
+Some OCP components currently use the `Always` image pull policy during
+upgrades. As a result, the kubelet and CRI-O will try to contact the registry
+server, even if the image is already available in the local storage of the
+cluster. This blocks upgrades and should be avoided.
+
+Most of these OCP components have been changed in the past to avoid this use of
+the `Always` pull policy. Recently the OVN pre-puller has also been changed
+(see this [bug](https://issues.redhat.com/browse/OCPBUGS-13219) for details).
+To prevent bugs like this happening in the future and make the boots and
+upgrades less fragile we should have a test that gates the OpenShift release
+and that verifies that boots and upgrades can be performed without a registry
+server. One way to ensure this is to run in CI an admission hook that
+rejects/warns about any spec that uses the `Always` pull policy.
+
+It would also be useful to have another test that scans for use of this
+`Always` pull policy in the source code.
+
+#### Don't try to contact the image registry server explicitly
+
+Some OCP components explicitly try to contact the registry server without a
+fallback alternative. These need to be changed so that they don't do it or so
+that they have a fallback mechanism when the registry server isn't available.
```

Comment:
some more examples:

the imagestream controller comes to mind (for imagestream import), as well as the build pods pull their own images because they don't have access to the node filesystem.

we need to think about/elaborate on how we expect each of those to behave....builds for example don't seem like a big deal.  imagestream import is trickier.

i think we want to have the exhaustive list of components that directly contact registries described in this EP so we can also discuss how each of them will need to tolerate this situation.

Reply:
Note that this registry accesses that block boots or upgrades, not registry accesses in general. For example, a build that fails doesn't block the upgrade, so I think it shouldn't be listed here.

I am afraid I can't provide an exhaustive list here, other than the ones that I already listed. The OVN upgrade pod has already been fixed, so I removed it. The MCO issue still exists, at least in 4.13.

The key point is that new ones can be introduced: that is why we are suggesting to try to detect then with CI tests that gate the release.

Reply:
> Note that this is about registry accesses that block boots or upgrades, not registry accesses in general. For example, a build that fails doesn't block the upgrade, so I think it shouldn't be listed here.

if the expectation is "a registry is not available during bootstrap + upgrade but will be available later" that is definitely simpler.  Even more so if you are saying "the registry is available at those times but we want to avoid relying on it as much as possible(because it's slow or unreliable)" as opposed to "the registry is not available and will never be"

In the first case(or third case) we still need to ensure we know which components will be trying to talk to a registry and what the implication is when the registry isn't available during those times.  In the second case obviously it's irrelevant.

Reply:
The most typical one is that access to the registry is possible but via a slow and/or unreliable connection. But we want to do this in way that will also help us when the registry server isn't and will never be available.

We have already explored what components try to talk to the registry during upgrades. We found the OVN upgrade, which we fixed, and we also found that MCO tries to fetch things directly (with no fallback since 4.13). There may be others, like the image streams that you mention. I don't really know, that is why I suggest that we create CI tests to find them.